pallets/permissions/src/lib.rs

```diff
@@ -0,0 +1,249 @@
+// Copyright 2021 Centrifuge Foundation (centrifuge.io).
+//
+// This file is part of the Centrifuge chain project.
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+///! A crate that defines a simple permissions logic.
+///! Users of the create must implement the `Properties` trait on a
+///! type of their choosing in order to use this pallet properly.
+pub use pallet::*;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+use frame_support::{dispatch::DispatchResult, pallet_prelude::*};
+use frame_system::pallet_prelude::*;
+
+pub trait Permissions<AccountId> {
```

Comment:
Something I've noticed about this API - right now it is up to the calling pallet to ensure the origin is an admin. I'd suggest we take the origin as a parameter and do that checking here, so that it's centralized.

If we want pallets to be able to invoke add/rm without the admin origin, we can potentially allow `root` as an admin origin for everything. This would also allow democracy to make changes in the case that the admin account is lost.

Reply:
Sorry, in this case I mean the _pool_ admin, not the democracy/root "AdminOrigin"

Reply:
The other option would be to move `approve_role_for` and `revoke_role_for` into the permissions pallet, then the API here only needs to be checking if an account is authorized?

The overall goal in my mind is
1) make it foolproof for other pallets to interact with this one, by not relying on external code to do permissions checks
2) ensure we have a council/democracy safety valve in case a pool admin loses their key

Reply:
I am not sure I follow your idea. 

* Do you want the trait to contain a caller, i.e. the origin, that this trait method is called from? 

I would like to not do this. My idea was, that the pallet-permissions provides root (i.e. the `AdminOrigin`) access to the storage. In this way root (if we set the `AdminOrigin` correctly) will always be able to fix permissions. 

Other pallets using the `trait Permissions` should not have to deal with carrying around an Origin in order to call this. I think it is fair to assume, that pallets should be able to just add and remove permissions from `AccountId`s.

Reply:
Regarding the `approve_role_for` and the `revoke_role_for` I think we can remove them. I just did not wanted to remove as much of the API as I was not sure about the reasoning here. 
I guess in case of pool-creation, the pallet-pool should take care of inserting an `PoolAdmin` into the permissions pallet.